nixos/tests/binary-cache.nix

```
@@ -0,0 +1,31 @@
+import ./make-test-python.nix ({ lib, ... }:
+
+with lib;
+
+{
+  name = "binary-cache";
+  meta.maintainers = with maintainers; [ thomasjm ];
+
+  nodes.machine =
+    { pkgs, ... }: {
+      imports = [ ../modules/installer/cd-dvd/channel.nix ];
+      environment.systemPackages = with pkgs; [hello python3];
+      system.extraDependencies = with pkgs; [hello.inputDerivation];
+      nix.extraOptions = ''
+        experimental-features = nix-command
+      '';
+    };
+
+  testScript = ''
+    machine.succeed("nix-build -E 'with import <nixpkgs> {}; mkBinaryCache { rootPaths = [hello]; }' -o /tmp/cache")
```

These assertions don't test the validity of the cache yet. If you remove `hello` from the `systemPackages`, something like this should show that the cache works:

```suggestion
    log = machine.succeed("nix-build -A hello '<nixpkgs>' --option require-sigs false --option trusted-substituters file://${mkBinaryCache {rootPaths = [hello];}} 2>&1")
    # some assert about log, to cover the possibility that hello may already be in the store by mistake
```

---

`hello.inputDerivation` will be unnecessary. We don't really need to build anything to prove that the cache works.

---

It does not work without `hello.inputDerivation` because some dependencies aren't present, and it tries to hit a cache over the network and fails. I was able to remove `hello` though.

I've improved the tests in a couple ways in the latest commits.

If we want to have a really solid test, I think the thing to do is
1) Start the VM and build the cache
2) Remove all paths like `/nix/store/...hello-...` from the store. (I wasn't able to figure out how to remove these since they're GC roots of the NixOS system, do you know how?)
3) Build using the cache and ensure a `/nix/store/...hello-...` path now exists.

---

Okay, I've figured out how to make a nice robust test! This is ready for another look now.